### PR TITLE
Restore clang tidy test.

### PR DESCRIPTION
**Pull request by trivialfis**
Close https://github.com/dmlc/xgboost/issues/8854 .

- Upgrade clang-tidy to 15.
- Fix clang-tidy warnings.

~One caveat of upgrading the clang version is that it might require a curand header.~

### CODE REVIEW

File: src/common/charconv.cc [hunk 1]
```diff
@@ -911,7 +911,7 @@ from_chars_result FromCharFloatImpl(const char *buffer, const int len,
   // the bias and also special-case the value 0.
   int32_t shift = (f_e2 == 0 ? 1 : f_e2) - exp_b2 - IEEE754::kFloatBias -
                   IEEE754::kFloatMantissaBits;
-  assert(shift >= 0);
+  assert(shift >= 1);
```
Comment (hcho3) [inline src/common/charconv.cc:914]: Why the change of condition here?

Comment (trivialfis) [reply src/common/charconv.cc:914]: Tell clang-tidy that `shift` must be greater or equal to zero. I have verified the condition using `CHECK` and the `RoundTripExhaustive` JSON test.

Comment (hcho3) [reply src/common/charconv.cc:914]: > Tell clang-tidy that shift must be greater or equal to zero.

But the new condition says `shift >= 1`?

Comment (trivialfis) [reply src/common/charconv.cc:914]: I will just put a nolint there.
